File: Python/Tools.py
import math

################################################
"""
check if specific number is prime
this needs fixed
"""


def is_prime(num):
    if num == 1:
        return False
    if num > 2:
        for x in range(2, math.ceil(math.sqrt(num)) + 1):
            if(num % x == 0):
                return False
    return True


################################################
"""
find sum of all factors, 1 included, self excluded
"""


def sum_factors(num):
    sum = 1
    #for squares
    if math.sqrt(num) % 1 == 0 and num != 1:
        sum += math.sqrt(num)
        for x in range(2, int(math.sqrt(num)) - 1):
            if num % x == 0:
                sum += x + (num / x)

    # for non-squares
    else:
        for x in range(2, math.ceil(math.sqrt(num))):
            if num % x == 0:
                sum += x + (num / x)
    return sum


################################################
"""
checks if string is pandigital within passed range
"""


def is_pandigital(num, low_lim, up_lim):
    for x in range(low_lim, up_lim + 1):
        if(str(x) not in num
        or len(num) != up_lim - low_lim +1):
            return False
    return True


################################################
"""
factorial
"""


def factorial(num):
    if num == 0:
        return 1
    for x in range(2, num):
        num *= x

    return num


################################################
"""
Sieve of Eratosthenes
Works well up to at least 10M
"""


def sieve_of_e(num):
    primes = []
    prime = [True for i in range(num + 1)]
    p = 2
    while p * p <= num:

        # If prime[p] is not changed, then it is a prime
        if prime[p]:

            # Update all multiples of p
            for i in range(p * 2, num + 1, p):
                prime[i] = False
        p += 1

    # Print all prime numbers
    for p in range(2, num):
        if prime[p]:
            primes.append(p)

    return primes

################################################
# checks is string is a palindrome


def is_palindrome(num):
    if num == num[::-1]:
        return True
    return False


################################################
"""
Sieve of Atkin
This code is contributed by Smitha
not working
"""


def sieve(limit):
    primes = []

    # 2 and 3 are known to be prime
    if limit > 2:
        primes.append(2)
    if limit > 3:
        primes.append(3)

    # Initialise the sieve array with False values
    sieve = [False] * limit
    for i in range(0, limit):
        sieve[i] = False

    x = 1
    while x * x < limit:
        y = 1
        while y * y < limit:

            # Main part of Sieve of Atkin
            n = (4 * x * x) + (y * y)
            if (n <= limit and (n % 12 == 1 or
                                n % 12 == 5)):
                sieve[n] ^= True

            n = (3 * x * x) + (y * y)
            if n <= limit and n % 12 == 7:
                sieve[n] ^= True

            n = (3 * x * x) - (y * y)
            if (x > y and n <= limit and
                    n % 12 == 11):
                sieve[n] ^= True
            y += 1
        x += 1

    # Mark all multiples of squares as non-prime
    r = 5
    while r * r < limit:
        if sieve[r]:
            for i in range(r * r, limit, r * r):
                sieve[i] = False

    # populate array
    for a in range(5, limit):
        if sieve[a]:
            primes.append(a)

    return primes

File: Java/src/Tools.java
import java.util.ArrayList;


public class Tools {
	
	public static void main(String[] args){
		
		
		
	}
	
	/***************************************************************************
	 * check if specific number is pan-digital within passed range
	 */
	public static boolean isPandigital(String digits, int lowLim, int upLim) {
		
		for (int l = lowLim; l <= upLim; l++) {			
			if(!digits.contains(String.valueOf(l)) 
			 || digits.length() != upLim - lowLim + 1){
				return false;
			}
		}
		return true;
	}
	
	/***************************************************************************
	 * check if specific number is prime
	 */
	public static boolean isPrime(int num){
		
		if(num == 1){
			return false;
		}
		for (int i = 2; i <= Math.ceil(Math.sqrt(num)); i++) {
			if(num % i == 0 && num != 2){
				return false;
			}
		}
		return true;
	}
	
	/***************************************************************************
	 * rotate int
	 */
	
	
	public static int[] rotate(int number) {
		int[] numbers = new int[String.valueOf(number).length()];
		
    	int start = number;
    	int numdigits = (int) Math.log10((double)number);
    	int multiplier = (int) Math.pow(10.0, (double)numdigits);
    	int ref = 0;
    	
    	while(true){
    		int r = number % 10;
    		number = number / 10;
    		number = number + multiplier * r;
    		numbers[ref] = number;
    		ref++;
    		if(number == start)
    			break;
    	}
    	return numbers;
	}
	
	/***************************************************************************
	 * decimal to binary
	 */
	public static String toBinary(int num){
	    	
		String bin = "";
        int[] binaryNum = new int[1000];
        int i = 0;
        
        while (num > 0){
            binaryNum[i] = num % 2;
            num = num / 2;
            i++;
        }
        for (int j = i - 1; j >= 0; j--){
            bin = bin + String.valueOf(binaryNum[j]);
        }
        return bin;
    }
	
	/***************************************************************************
	 * check if integer is a palindrome
	 */
	public static boolean isPalindrome(int num){
		
		boolean isPalindrome = true;
		String temp = Integer.toString(num);
		int[] iBroken = new int[temp.length()];
		int[] iBackwards = new int[temp.length()];
		for (int a = 0, b = temp.length() - 1; a < temp.length(); a++, b--)
		{
			iBroken[a] = temp.charAt(a) - '0';
			iBackwards[b] = temp.charAt(a) - '0';
		}
		for (int a = 0; a < temp.length(); a++) {
			if(iBroken[a] != iBackwards[a]){
				isPalindrome = false;
			}
		}
		return isPalindrome;	
	}
	

	/***************************************************************************
	 * check if string is a palindrome
	 */
	public static boolean isStrPalindrome(String str){
		
		boolean isPalindrome = true;
		
		char[] iBroken = new char[str.length()];
		char[] iBackwards = new char[str.length()];
		for (int a = 0, b = str.length() - 1; a < str.length(); a++, b--)
		{
			iBroken[a] = (char) (str.charAt(a) - '0');
			iBackwards[b] = (char) (str.charAt(a) - '0');
		}
		for (int a = 0; a < str.length(); a++) {
			if(iBroken[a] != iBackwards[a]){
				isPalindrome = false;
			}
		}
		return isPalindrome;	
		
		/*
		String rev = new StringBuilder(str).reverse().toString();
		return str == rev;	
		*/
		
	}
	
	/***************************************************************************
	 * sieve of Eratosthenes
	 * This code has been contributed by Amit Khandelwal. 
	 * 
	 * this only works up to n = 7691
	 */
	
	 public static void sieveOfErt(int n) { 
	      
		 boolean prime[] = new boolean[n+1]; 
		 for(int i=0;i<n;i++) {
	            prime[i] = true; 
		 }
	          
		 for(int p = 2; p*p <=n; p++) { 
			 // If prime[p] is not changed, then it is a prime 
			 if(prime[p] == true){ 
				 // Update all multiples of p 
				 for(int i = p*2; i <= n; i += p){
	                    prime[i] = false;
				 }
			 } 
		 }
		 for(int i = 2; i <= n; i++){ 
			 if(prime[i] == true){
	                System.out.print(i + " "); 
			 }
		 } 
	 }
	
	/***************************************************************************
	 * sieve of Atkin
	 * This code is contributed by Anant Agarwal. 
	 */
static ArrayList<Integer> primes = new ArrayList<Integer>();
	
    public static ArrayList<Integer> sieve(int limit){
    	
    	if (limit > 2){ primes.add(2);}
    	if (limit > 3){ primes.add(3);}
    	  
    	boolean sieve[] = new boolean[limit]; 
    	  
    	for (int i = 0; i < limit; i++){
    		sieve[i] = false; 
    	}
    	for (int x = 1; x * x < limit; x++) { 
    		for (int y = 1; y * y < limit; y++) { 
    			int n = (4 * x * x) + (y * y); 
    			if (n <= limit && (n % 12 == 1 || n % 12 == 5)){
    				sieve[n] ^= true; 
    			}
    			n = (3 * x * x) + (y * y); 
    			if (n <= limit && n % 12 == 7){
    				sieve[n] ^= true; 
    			}
    			n = (3 * x * x) - (y * y); 
    			if (x > y && n <= limit && n % 12 == 11){ 
    				sieve[n] ^= true; 
    			}
    		} 
    	} 
    	  
    	for (int r = 5; r * r < limit; r++) { 
    		if (sieve[r]) { 
    			for (int i = r * r; i < limit; i += r * r){ 
    				sieve[i] = false; 
    			}
    		} 
    	} 
    	  
    	for (int a = 5; a < limit; a++){ 
    		if (sieve[a]){
    			primes.add(a);
    		}
    	}
		return primes;
    }
	
}
